cpp/include/cudf/utilities/default_stream.hpp

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime_api.h>
+
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+
+namespace detail {
+#if __cpp_inline_variables
+inline cudaStream_t default_stream;
+#else
+extern cudaStream_t default_stream;
+#endif
+}  // namespace detail
+
+/**
+ * @brief Set the default CUDA stream. This can only be done once.
+ *
+ * @param stream The stream to set as the default. Only `cudaStreamLegacy` or `cudaStreamPerThread`
+ * is allowed.
+ */
+void set_default_stream(cudaStream_t stream)
+{
+  CUDF_EXPECTS(detail::default_stream == cudaStreamDefault, "Default stream can only be set once");
+  CUDF_EXPECTS(stream == cudaStreamLegacy || stream == cudaStreamPerThread,
+               "Default stream can only be set to legacy or per-thread");
+  detail::default_stream = stream;
+}
+
+/**
+ * @brief Get the default CUDA stream.
+ *
+ * @returns cudaStream_t The default CUDA stream.
+ */
+cudaStream_t get_default_stream() { return detail::default_stream; }
```

Comment:
I think this should be made thread safe, and probably use a myers singleton like we had for `get_default_resource` (replaced by get_current_device_resource).

Also, this isn't safe for multidevice -- a user could set a non-default stream associated with device 0, which would then cause errors to use on device 1. So it probably should be `per_device_stream`...

Reply:
Wouldn't making it thread safe introduce a lot of overhead? At least in Spark's use case, we are going to set the default stream once during startup and never change it again. I imagine most use cases would be similar. Seems silly to change the default stream all the time.

The setter only allows `cudaStreamLegacy` or `cudaStreamPerThread`, so multiple devices shouldn't be an issue.